### ISSUE TITLE
Train a fast-rcnn CovNet on another dataset

### ISSUE THREAD

**Issue by SunShineMin** [rbgirshick/fast-rcnn#11]
Hi, I am trying to train a fast-rcnn CovNet on another dataset ,and there is something wrong with the snapshot when it comes to 
net.params['bbox_pred'][0].data[...] =  (net.params['bbox_pred'][0].data *self.bbox_stds[:, np.newaxis])
ValueError: operands could not be broadcast together with shapes (84,4096) (12,1) .
I think that maybe the difference of the number of classes between my dataset and VOC is account for the issues . If is it necessary to chang some parameters of the pretrained_model? If it is ,how should I change that?

**Comment by rbgirshick** [comment]
It looks like you have 3 classes. In the train.prototxt and test.prototxt files that you're using, you'll need to change `num_output` from 21 to 3 in the `cls_score` layer and from 84 to 12 in the `bbox_pred` layer. You'll also need to change `num_classes` from 21 to 3 in the Python layer that provides data to the net (the very first layer).

**Comment by SunShineMin** [comment]
Thank you very much for your reply and I have solved the problem with your help.But  when I increased the number of classes to 12 ,an AssertionError occured in imdb.append_flipped_images----assert (boxes[:, 2] >= boxes[:, 0]).all() .Is there something wrong with my proposals? But I have used these proposals in RCNN successfully.

**Comment by nascimentocrafael** [comment]
Hi SunShineMin,

Could you please share the steps to train fast-rcnn ConvNet on another dataset? Also, I'm having problems downloading the PASCAL dataset, it seems that the links are broken. Did you have some problem downloading the database?. If you help me I'll be extremely thankful. If you prefer, you could e-mail me.

Thank you in advance.

**Comment by ssakhavi** [comment]
@SunShineMin @rbgirshick 
Some of the forks from this repo are actually trying (or succeeded) to do this. 

I refer you to:
[https://github.com/EdisonResearch/fast-rcnn](https://github.com/EdisonResearch/fast-rcnn)
[https://github.com/raingo/fast-rcnn](https://github.com/raingo/fast-rcnn)

**Comment by zeyuanxy** [comment]
@SunShineMin @rbgirshick @nascimentocrafael  @ssakhavi  I have succeeded training it on INRIA Person, see https://github.com/EdisonResearch/fast-rcnn  for more details. Thanks!

**Comment by zeyuanxy** [comment]
How to Train Fast-RCNN on Another Dataset

https://github.com/EdisonResearch/fast-rcnn/tree/master/help/train

**Comment by SunShineMin** [comment]
@ssakhavi @zeyuanxy @rbgirshick Thank you very much for above help. I have trained Fast R-CNN on my own dataset, but the mAP is lower than R-CNN . I changed the learning rate and lterations, with little improvement . Which arguments can I modify else to improve mAP ?

**Comment by SunShineMin** [comment]
@nascimentocrafael  I have sent you an email attached with VOC dataset , please have a check

**Comment by zeyuanxy** [comment]
@SunShineMin  Which dataset did you train Fast-RCNN on? I am so glad to meet a Tsinghua peer here.

**Comment by SunShineMin** [comment]
So am I. I made a new dataset for my research. Perhaps we can have a communication for detail on QQ: 1114784152

**Comment by github-anurag** [comment]
@zeyuanxy 
I am having a bit of trouble while running the steps mentioned in 
https://github.com/EdisonResearch/fast-rcnn/tree/master/help/train 

When I attempt to train the model I get the following error

Traceback (most recent call last):
  File "./tools/train_net.py", line 78, in <module>
    imdb = get_imdb(args.imdb_name)
  File "/Users/anurag/fast/fast-rcnn/tools/../lib/datasets/factory.py", line 44, in get_imdb
    return __sets[name]()
  File "/Users/anurag/fast/fast-rcnn/tools/../lib/datasets/factory.py", line 53, in <lambda>
    __sets[name] = (lambda split=split: datasets.inria(split, inria_devkit_path))
TypeError: 'module' object is not callable

Its most probably related to a import but I cant figure out which one.

**Comment by github-anurag** [comment]
@rbgirshick @zeyuanxy @SunShineMin 

How do we go about training for the negative/background class? After reading the paper, I understand that if the IoU between the bbox proposals and our labelled bbox is < 0.3 we consider those bboxes to be negative. Do we specify these bboxes as background class for each image while creating the training set imdb's? 

I imagine a caffe layer can do this .. does something like that exist already? thanks

**Comment by hengck23** [comment]
@ github-anurag 
If you look at " fast-rcnn/lib/datasets/pascal_voc.py",
              def selective_search_roidb(self):
                   if int(self._year) == 2007 or self._image_set != 'test':
line 124:            gt_roidb = self.gt_roidb()
line 125:            ss_roidb = self._load_selective_search_roidb(gt_roidb)
line 126:            roidb = datasets.imdb.merge_roidbs(gt_roidb, ss_roidb)

It means that during training, the ground truth annotation is first read in line 124. These are positive samples. Then selective search boxes are loaded, those that do not overlap with the ground truth are marked as backgrounds (aka negative samples), as in line 125. Finally, both positive and negative samples are combined into roidb and used for training in line 126.

note that "selective_search_roidb()" is always called when you use imdb.roidb(), see
" fast-rcnn/lib/datasets/imdb.py:" 
    @property
    def roidb(self):
imdb is your train database and roidb is the roi of your train database.

in short, you need to supply ground truth annotation and a set of box list (e.g. from selective search) for training. Hope this explanation helps.

**Comment by anuchandra** [comment]
@zeyuanxy @SunShineMin

Thanks for the instructions on how to train on your own dataset. I tried to follow them using one imagenet class and a background. But I'm getting an error which is exactly the same as SunShineMin did.

As background, I computed the selective search proposals. Then ran the train_net.py script. Any suggestions? SunShineMin, how did you fix this? Anyone else have any suggestions?

Traceback (most recent call last):
  File "./tools/train_net.py", line 80, in <module>
    roidb = get_training_roidb(imdb)
  File "/home/ubuntu/fast-rcnn/tools/../lib/fast_rcnn/train.py", line 107, in get_training_roidb
    imdb.append_flipped_images()
  File "/home/ubuntu/fast-rcnn/tools/../lib/datasets/imdb.py", line 104, in append_flipped_images
    assert (boxes[:, 2] >= boxes[:, 0]).all()
AssertionError
ubuntu@ip-172-31-8-77:~/fast-rcnn$

**Comment by sherifshehata** [comment]
Hi all,
I changed the code to work on my dataset according to the comments by @rbgirshick and the help page by @zeyuanxy (thanks for the much helpful info). The code loads the data correctly and without problems, but then i get an error in the first training iteration.

This is the error:
I0717 10:25:23.420873  2785 solver.cpp:189] Iteration 0, loss = 2.75509
I0717 10:25:23.420953  2785 solver.cpp:204]     Train net output #0: loss_bbox = 0.212742 (* 1 = 0.212742 loss)
I0717 10:25:23.421013  2785 solver.cpp:204]     Train net output #1: loss_cls = 2.54235 (* 1 = 2.54235 loss)
I0717 10:25:23.421068  2785 solver.cpp:464] Iteration 0, lr = 0.001
F0717 10:25:23.985546  2785 roi_pooling_layer.cu:91] Check failed: error == cudaSuccess (9 vs. 0)  invalid configuration argument

According to caffe users comments i tried different batch_size (even tried batch_size = 1), but it didn't solve the problem.

Any clue what causes this problem?

**Comment by github-anurag** [comment]
@anuchandra 
This assertion error is due to some problems with your annotation files. The annotations have the bounding box informations as column 0 - xmin, column 1- ymin , column 2- xmax, column 3 as ymax. The assertion fails as you have xmin > xmax for one or possibly more of your annotation files. Try to fix this.

**Comment by YihangLou** [comment]
@anuchandra 
I know what caused this error. Because I have met this error too and the solved. The reason is probably due to the "-1" operation during reading the anontation file or your selectivesearchBox.mat file. You should notice your pixel coordinates are whether starting from 0 or 1, if you coordiantes is starting from 0, do not perform "-1" operation which you can see in the pascal_voc.py. And this operation you can find in function _load_pascal_annotation() and _load_selective_search_roidb. :)

**Comment by anuchandra** [comment]
@YihangLou 

You're absolutely right. I added the checks for negative x1 and y1 in _load_pascal_annotation and this fixed a couple of issues. Firstly, the assert (boxes[:, 2] >= boxes[:, 0]).all() were no longer called. The second big thing was I was getting nans appearing in my loss_bbox during training. These have gone as well. I carefully compared the annotations which were being made negative x1s with those caught by the assert. It seems the assert wasn't catching all of them. Thanks!

**Comment by rockrush** [comment]
I trained on custom dataset under GPU mode(Titan), and got "nan" in result_vec[k](solver.cpp) all the way, what can I do then?

**Comment by sunshineatnoon** [comment]
@hengck23 If I understand correctly, you mean fast rcnn will generate negative training data for us, so the only thing we need to provide is the ground truth bounding box?

**Comment by sunshineatnoon** [comment]
@anuchandra Have you solved the assertion problem? I trained on image net and have the same problem, by deleting -1 operation, I still have the same problem.

**Comment by anuchandra** [comment]
@sunshineatnoon I kept the -1 operations. Instead I added checks in _load_pascal_annotation() 
In_load_pascal_annotation(), after these lines, 
            # Make pixel indexes 0-based
            x1 = float(get_data_from_tag(obj, 'xmin')) - 1
            y1 = float(get_data_from_tag(obj, 'ymin')) - 1
            x2 = float(get_data_from_tag(obj, 'xmax')) - 1
            y2 = float(get_data_from_tag(obj, 'ymax')) - 1

I added:
            if x1 < 0:
				x1 = 0
            if y1 < 0:
				y1 = 0
That took care of annotations where x1 or y1 were 0.

**Comment by sunshineatnoon** [comment]
@anuchandra Thanks for you help. I added those lines too, but I still get this assert error, any thoughts?

**Comment by anuchandra** [comment]
@sunshineatnoon Similarly in addition, you could try checking the -1 operation in _load_selective_search_roidb is doing. See @YihangLou suggestion above. This didn't prove an issue for my dataset but it could be with yours.

     box_list.append(raw_data[i][:, (1, 0, 3, 2)] - 1)

**Comment by IchibanKanobee** [comment]
I have a couple of questions about training the fast r-cnn on another dataset, as it is described in
https://github.com/EdisonResearch/fast-rcnn/tree/master/help/train

Do I put negative and positive images under the same "INRIA/data/Images" folder, or separate them into sub-folders, like "INRIA/data/Images/positive" and "INRIA/data/Images/negative". The more general question is, if I have several categories, do I put all of their images into the same "INRIA/data/Images" folder, or into separate  "INRIA/data/Images/Category1"... "INRIA/data/Images/CategoryN" sub-folders?

The second question is, do I need annotations for negative samples as well, or only for positive ones?

Thanks a lot for your help!

**Comment by sid027** [comment]
@IchibanKanobee were you able to figure out if one has to create subfolders?

**Comment by IchibanKanobee** [comment]
I am still not able to train on my own data set, so I might be wrong, but my understanding is that it is not necessary to separate the images into sub folders. At least the selective_search.py is written in a way that all images are under the same folder for generating rois.
I also understand that annotations are needed for negative samples as well, but I am not sure if the region of interest in this case is the whole image with the center pointing to the center of the image, or it is [0, 0, 0, 0] with the center pointing to [0,0].

**Comment by catsdogone** [comment]
@zeyuanxy 
I try to train on my database as it is describe in https://github.com/EdisonResearch/fast-rcnn/tree/master/help/train, while there is something wrong:
... 
 File "/usr/local/fast-rcnn/tools/../lib/datasets/imdb.py", line 167, in create_roidb_from_box_list
    argmaxes = gt_overlaps.argmax(axis=1)
ValueError: attempt to get argmax of an empty sequence

Thanks a lot for your help!

**Comment by sid027** [comment]
@IchibanKanobee I think it is the corner.I think annotations are needed for all classes except the background.I am trying to train the detector for the ILSVRC2012 imagenet dataset.Did you use the same dataset?

**Comment by zeyuanxy** [comment]
@catsdogone  It seems that there is some problem with your selective search file.

**Comment by IchibanKanobee** [comment]
@sid027 I am following the training instructions in readme.md in help folder and using the VGG_CNN_M_1024 model. 

If I don't have annotations for background files, I am getting an error:
File "./tools/train_net.py", line 80, in <module>
    roidb = get_training_roidb(imdb)
  File "/home/GitHub/fast-rcnn/tools/../lib/fast_rcnn/train.py", line 111, in get_training_roidb
    rdl_roidb.prepare_roidb(imdb)
  File "/home/GitHub/fast-rcnn/tools/../lib/roi_data_layer/roidb.py", line 23, in prepare_roidb
    roidb[i]['image'] = imdb.image_path_at(i)
IndexError: list index out of range

That's why I thought the annotations for background files are required. But I didn't get the training working, so I might be wrong.

**Comment by catsdogone** [comment]
@zeyuanxy @rbgirshick 
I use ‘selective_search' to produce region proposals for '/fast-rcnn/data/demo/000004.jpg'(get 2000+boxes), while the detection result is bad compare to using the '000004_boxes.mat' which is provided. What I can do to get a better result for myself image data.
Thanks for your help.

**Comment by IchibanKanobee** [comment]
@zeyuanxy Do you include background files in the training set? If they are included in the training set, do you include annotations for the background files? And, if you do include annotations for the background files, what are the bounding boxes - the whole image, or [0,0,0, 0]?

Thanks!

**Comment by anuchandra** [comment]
@IchibanKanobee The easiest way to figure out how fastrcnn works is to take the working PASCAL VOC 2007 example and strip it down to one image in one class. You can use the Edison training guide - it'll help you do that.

**Comment by IchibanKanobee** [comment]
@anuchandra Thank you for the tip! Can you please point me to the working PASCAL VOC 2007 sample?

**Comment by anuchandra** [comment]
The instructions for the working example are on the home page.

https://github.com/rbgirshick/fast-rcnn#beyond-the-demo-installation-for-training-and-testing-models

All the info you need is on these discussion forums.

**Comment by sunshineatnoon** [comment]
@zeyuanxy Hi~ Thanks to your INRIA docs, I successfully train a model on INRIA. I don't have matlab on my computer, so I use dlib's selective search. I want to create a version that can be ran without matlab. So I need to wrtire an evaluation script in python. But I don't know how to evaluate prediction precision on INRIA. Could you pls give me some hints? Thanks.

**Comment by catsdogone** [comment]
@sunshineatnoon 
Maybe this is what you need：
https://github.com/rbgirshick/fast-rcnn/pull/33/files

**Comment by sunshineatnoon** [comment]
@anuchandra Thx, getting rid of the -1 operation solves my problem.

**Comment by sunshineatnoon** [comment]
@catsdogone Thx, I wrote some test code myself.

**Comment by yawadugyamfi** [comment]
When passing argument for training the INRIA person dataset what does inria_train represent in the code below. Is it the train.mat file generated from selective search? or the folder containing images of your training data?

./tools/train_net.py --gpu 0 --solver models/VGG_CNN_M_1024/solver.prototxt \
    --weights data/imagenet_models/VGG_CNN_M_1024.v2.caffemodel --imdb inria_train

**Comment by zsc-tju** [comment]
@zeyuanxy @rbgirshick 
I have a question,now that we set the scale=600 in the config.py, why in the roi_data_layer\layer.py,still to reshape the top[0] to 100x100,and in the test phase the iunput size is 227x227,please enlighten me,thank you very much!

**Comment by MinaRe** [comment]
Hey Guys ,
I have same question and I have found this link before 
https://github.com/zeyuanxy/fast-rcnn/tree/master/help/train
, Actually I dont know How should prepare Annotation.txt and annotation file? 
I'd appreciate any comments :)

**Comment by siddharthm83** [comment]
@anuchandra @SunShineMin @sunshineatnoon 
I was working on the imagenet dataset and I got the assertion error  in imdb.append_flipped_images----assert (boxes[:, 2] >= boxes[:, 0]).all() 
From: http://image-net.org/download-bboxes
`Remark: In the bounding box annotations, there are two fields(<width> and <height>) indicating the size of the image. The location and size of a bounding box in the annotation file are relative to this size. However, this size may not be identical to the real image size in the downloaded package. (The reason is that the size in the annotation file is the displayed size in which the image was shown to an annotator). Therefore to locate the actual pixels on the original image, you might need to rescale the bounding boxes accordingly.`

It appears that one would need to normalize values in xml to the true size of the image.

**Comment by attiliotnt** [comment]
@IchibanKanobee  how did you fixed the "IndexError: list index out of range"?
I'm stucked there since two days trying to train my own dataset.

**Comment by IchibanKanobee** [comment]
@attiliotnt Make sure to delete the old cache folder before starting the new training

**Comment by attiliotnt** [comment]
Thank you for the answer @IchibanKanobee.
My problem was that the annotation files I got from Matlab didn't have the fields _folder_ and _name_

**Comment by anuchandra** [comment]
@siddharthm83 This is useful to know for the future.

**Comment by MinaRe** [comment]
Dear @catsdogone and @zeyuanxy 

I have same problem in roidb.py ,is this problem because of train.mat???
*******__init__
    rdl_roidb.add_bbox_regression_targets(roidb)
  File "/data/Workspace/fast-rcnn/tools/../lib/roi_data_layer/roidb.py", line 57, in add_bbox_regression_targets
    max_overlaps = roidb[im_i]['max_overlaps']
KeyError: 'max_overlaps'****************

**Comment by MinaRe** [comment]
Dear All
finaly i run fast-rcnn on my dataset and get acceptable results on object detection task, but I was wondering to know about the segmentation task? has anyone done it before?

**Comment by attiliotnt** [comment]
Hello, I would like to ask which script you chose to run for detection? I tried to use my webcam but the results are not so good.

**Comment by MinaRe** [comment]
I apply fast-rcnn for cancer cell detection.

**Comment by attiliotnt** [comment]
@MinaRe  very interesting. Do you use images or video for test?

**Comment by MinaRe** [comment]
3D images :)
 do you know about the segmentation task by fast rcnn?

**Comment by attiliotnt** [comment]
@MinaRe  No, unfortunately. I worked just with detection.

**Comment by attiliotnt** [comment]
I would like to ask an advice about the model to use during the training. I need to detect some object underwater and I used CaffeNet.v2.caffemodel. The result I obtained are really bad until now. Any advice?

**Comment by MinaRe** [comment]
I did training on Asus-G751JT on CPU.

**Comment by attiliotnt** [comment]
I have a little doubt. During the labeling of the training set, is it possible to label more times items (ROIs) in the same image?

**Comment by IchibanKanobee** [comment]
@attiliotnt Yes, it is. If you want recognize facial features, for example, you specify eyes, mouth, nose, etc. on the same image.

**Comment by attiliotnt** [comment]
@IchibanKanobee  in my case I have to recognize just a kind of items (2 classes including the background) but in the training set there are several images containing more than a single item in the same image

**Comment by attiliotnt** [comment]
@SunShineMin  @anuchandra @siddharthm83   I'm having the assertion error too.
I inserted the check after _load_annotation() but it's the same.
If I don't remove -1 in

box_list.append(raw_data[i][:, (1, 0, 3, 2)] - 1)

I get `assert all(max_classes[nonzero_inds] != 0)`

I don't know how to fix this

**Comment by attiliotnt** [comment]
@IchibanKanobee  thank you for the explanation!

**Comment by attiliotnt** [comment]
Could anybody give me a suggestion about how to get rid of AssertionError?

**Comment by anuchandra** [comment]
@attiliotnt I can give you a generic suggestion. When I run into a tough problem that no one else can seem to help with, I break it down. The way I got fast-rcnn building a model was to take the working example - see repo frontpage ie VOC2007 datasets. Stripped them down from 20 classes to 1 class. Stripped the one class from tons of images to 1 image. My first fast-rcnn model was 1 class with 1 image. Subsequently, I added images in small batches. This approach meant that when I had an error like you do, I could pinpoint down to a single image.

**Comment by attiliotnt** [comment]
Thanks @anuchandra for the suggestion. Actually I was trying to improve my previous dataset adding more pictures (I only use two classes including the background). The error I got happened using pictures from Imagenet database and I temporarily solve it just don't using them.

**Comment by anuchandra** [comment]
@attiliotnt If you add imagenet images & annotations in progressively smaller batches, you should be able to identify which image or annotation file in the imagenet dataset is causing the assert. You can write scripts which speed it up. This is literally what I did at one point.

**Comment by deeha** [comment]
@MinaRe i really need to know , how you run it? i need steps.  kindly guide me. thanks alot

**Comment by MinaRe** [comment]
@deeha  I followed step by step of https://recordnotfound.com/fast-rcnn-coldmanck-53618 ,let me know if you have any problem after reading this page :)

**Comment by karaspd** [comment]
@zeyuanxy , @SunShineMin , @IchibanKanobee, @MinaRe I did training on INRIA dataset but when I do demo I saw lots of  little boxes with confidence of higher than 0.5 (and all same) at the bottom or side of demo images and some correct boxes on the object. Unfortunately, I am behind proxy and I cannot send any figures!! I was wondering if any of you see this issue previously. I was suspected of train scales however changing it to [300,400,500,600,700,800] did not solve it. Please let me know if you have any solution to this problem.

**Comment by karaspd** [comment]
I found my problem :) for future refrence, the problem was because of object proposal coming from matlab selective search and demo.py. selective search was given boxes in the format of (y1,x1,y2,x2) 
obj_proposal = sio.loadmat(box_file)['boxes']
I swapped x1 with y1 and x2 with y2.
        for i in range(len(obj_proposal)):
            obj_proposal[i,0], obj_proposal[i,1]=obj_proposal[i,1], obj_proposal[i,0]
            obj_proposal[i,2], obj_proposal[i,3]=obj_proposal[i,3], obj_proposal[i,2]

**Comment by deeha** [comment]
@MinaRe i have problem like this:

selective_search
Error using textread (line 165)
File not found.

Error in selective_search (line 14)
image_filenames = textread([image_db '/data/ImageSets/train.txt'], '%s', 'delimiter', '\n');

**Comment by attiliotnt** [comment]
@deeha  You can easily solve it leaving temporarily a copy of the train.txt in the Images folder.

**Comment by deeha** [comment]
@attiliotnt  still same error. i think error is "textread".

**Comment by attiliotnt** [comment]
Is there already any script for testing fast-rcnn on videos?

**Comment by mksarker** [comment]
hi all,
I have followed the below web instruction for training my own dataset as INRIA,
https://github.com/zeyuanxy/fast-rcnn/blob/master/help/train/README.md
and found the below problem,
:~$ cd fast-rcnn
:~/fast-rcnn$ ./tools/train_net.py --gpu 0 --solver models/pascal_voc/VGG_CNN_M_1024/fast-rcnn/solver.prototxt --weights data/faster_rcnn_models/VGG16_faster_rcnn_final.caffemodel --imdb inria_train
/home/sarker/anaconda2/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Python 2.7.11 |Anaconda custom (64-bit)| (default, Jun 15 2016, 15:21:30) 
Type "copyright", "credits" or "license" for more information.

IPython 4.2.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: 


when I run the training code on terminal the IPython console is open? I don’t know what is the problem? I am a beginner  in this area. Please help me.
Thanks in advanced....

**Comment by abhisheksgumadi** [comment]
I still get this error

argmaxes = gt_overlaps.argmax(axis=1)
ValueError: attempt to get argmax of an empty sequence

Any idea why this is happening? I am training on my own set of images and all annotation files look fine for me.

Thanks !

**Comment by anuchandra** [comment]
@abhisheksgumadi Most likely an issue with the annotation files. You should be able to print out which annotation file is causing this exception. I forget which bit of code causes this error but add a print somewhere which prints out the name of the annotation file. This should help you pinpoint the annotation file in question. The other way is to try and reduce the number of annotations to a small working set and progressively add annotations until you find the one causing the problem.

**Comment by abhisheksgumadi** [comment]
Thanks @anuchandra  I got the error. It was because I had few annotation files that did not have any object annotations (bounding boxes) in them. I eliminated them and now it runs fine.

Having solved that, I have another question. How should the annotation files look if I have to include images that should completely be treated as background images? Should I have a object annotation in the XML file with the bounding box covering the entire image and the name  of the object as  "__background__" ?

**Comment by anewlearner** [comment]
@MinaRe

Hello, MinaRe
I already know how to train on new 2d datasets.
What should I do if I want to apply fast-rcnn to 3d images? I am not sure if there is any other changes except for `train.prototxt` and `test.prototxt`. It would be better if you can provide an example.

Thanks.

**Comment by xinleipan** [comment]
What does " Write the function for parsing annotations, see _load_sgs_annotation in sgs.py" and "Do not forget to add import syntaxes in your own python file and other python files in the same directory" mean? 

Can any one give a step by step detailed instruction on how to train Faster-RCNN on customized dataset? Thanks!

**Comment by Terminator1500** [comment]
I am trying to train fast-RCNN on INRIA dataset following the instructions as given by @zeyuanxy  at https://github.com/zeyuanxy/fast-rcnn/blob/master/help/train/README.md. But I have not been able to achieve any appreciable results. The output of the trained fast-RCNN gives **all** the input selective search bounding boxes and **all with the same score**. Below is a brief of approach I followed:
I have prepared the 'inria.py' file, put the required code in factory.py. I have computed the selective search boxes using the code by EdisonResearch. Then I have modified Prototxt number of classes, and running train_net.py.
Can any one suggest me where the problem might be originating? It is quite odd and frustating to get every proposal's score to be exactly same everytime I train a model.
Thank You!

**Comment by micros-uav** [comment]
@attiliotnt I also have this error:
assert all(max_classes[nonzero_inds] != 0)
AssertionError

Have you found the solution?

**Comment by ruqiang826** [comment]
@Terminator1500 I have the same problem of same score. have you found the solution?

**Comment by Terminator1500** [comment]
@ruqiang826 Well, Earlier I was not using the pre-trained model on imagenet and training it from the scratch. Using pretrained imagenet models does improve the results a bit. See if it works for you.

**Comment by smajida** [comment]
@Terminator1500 I am also having the same problem of same scores for different objects. The results are strange as I get same detection bboxes for different classes. Have you found the solution?

**Comment by indsak** [comment]
Hi
i am trying to train faster RCNN as per https://github.com/deboc/py-faster-rcnn/tree/master/help
 
When i try to train using the command
./tools/train_faster_rcnn_alt_opt.py --gpu 0 --net_name fishclassify --weights data/imagenet_models/VGG_CNN_M_1024.v2.caffemodel --imdb fishclassify --cfg /home/alpha/Indhu/fast-rcnn/py-faster-rcnn/config.yml

I am getting the following error
File "/fast-rcnn/py-faster-rcnn/tools/../lib/datasets/factory.py", line 46, in get_imdb
    raise KeyError('Unknown dataset: {}'.format(name))

Any help on what might be causing this?
I had created fishclassify.py and fishclassify_eval.py under lib/datasets as well as modified factory.py

**Comment by Kimbrel** [comment]
Hi
I am trying to train fast RCNN as per  https://github.com/zeyuanxy/fast-rcnn/blob/master/help/train/README.md
I tried using the net created on the demo file, but it doeskin show any bounding boxes
Can anyone help on what might be causing this?
